soco/discovery.py

```diff
@@ -239,3 +242,125 @@ def by_name(name):
         if device.player_name == name:
             return device
     return None
+
+
+def scan_network(max_threads=256, timeout=3.0, include_invisible=False):
+    """Scan all attached networks for Sonos devices.
+
+    Scans the IPv4 network attached to each interface to check for network devices
+    with port 1400 open. Check IPs in parallel threads for efficiency. Once the first
+    Sonos device is found, stop checking and use that device to find the other
+    devices. Returns a set of `SoCo` instances, or `None` if no Sonos devices are
+    discovered.
+
+    This function is intended for use when the usual discovery functions aren't
+    working, perhaps due to multicast problems on the network to which the SoCo
+    host is attached.
+
+    Args:
+        max_threads (int): The maximum number of threads to use when scanning the
+            network.
+        timeout (float): The network timeout in seconds to use when checking
+            each IP address for a Sonos device
+        include_invisible (bool): Whether to include invisible Sonos devices in
+            the set of devices returned.
+
+    Returns:
+        set: a set of `SoCo` instances, one for each zone found, or else `None`.
+    """
+
+    def is_ipv4_address(ip_address):
+        """Helper function to test for an IPv4 address."""
+        try:
+            ipaddress.IPv4Network(ip_address)
+            return True
+        except ValueError:
+            return False
+
+    def find_ipv4_networks():
+        """Helper function to return a list of unique IPv4 networks to which
+        this node is attached.
+        """
+        ipv4_net_list = []
+        adapters = ifaddr.get_adapters()
+        for adapter in adapters:
+            for ip_address in adapter.ips:
+                if is_ipv4_address(ip_address.ip):
+                    # Restrict to common domestic private IP ranges and sensible
+                    # netmasks. Experimental ... assumptions need to be tested.
+                    if (
+                        ip_address.ip.startswith("192.168")
+                        or ip_address.ip.startswith("10.")
+                    ) and ip_address.network_prefix <= 24:
```

Comment:
This check would currently ignore the 172.16/12 network. Instead of hardcoded checks, you could use `ip_network()` and the `is_private` property (and `is_loopback` for 127/8):
```
network = ip_network("address")
if not network.is_private or network.is_loopback:
    continue
```

`supernet(new_prefix=24)` can be used to force the network prefix to a specific length, with the caveat of ignoring the fact that some may have larger home networks :-)

Reply:
Great -- thanks. This is much better. I'll try it out.

Reply:
Done. Thanks for the help.